enhancements/network/smart-nic-ovn-offload.md

```diff
@@ -0,0 +1,450 @@
+      limits:
+        openshift.io/<sriov-resource-name>: 1
+```
+
+In the above pod yaml file:
+
+`openshift.io/<sriov-resource-name>` is the full resource name that follows
+the [extended resource naming scheme](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources),
+it is advertised by SR-IOV Network Device Plugin and published in Kubernetes
+node status. The SR-IOV Network Device Plugin is a component that discovers
+and manages the limited SR-IOV devices on the SmartNIC host. The managed
+SR-IOV devices, AKA VFs, are created from SmartNIC PF.
+
+`1` is the number of SR-IOV devices requested for the `sriov-pod`. When the
+requested number of SR-IOV devices are not available on the target node, the
+`sriov-pod` will be in `Pending` state until resource become available again
+(e.g. other SR-IOV pods release the VF device on the target node).
+
+Given that regular **veth** pods cannot be created on the SmartNIC host,
+failure would happen if some cluster pods or infra pods using **veth**
+interfaces are scheduled on the SmartNIC host, examples of cluster or infra
+pods are as below:
+
+- ingress-canary in openshift-ingress-canary namespace
+- image-pruner in openshift-image-registry namespace
+- network-metrics-daemon in openshift-multus namespace
+- network-check-source & network-check-target in openshift-network-diagnostics
+- Pods in openshift-marketplace & openshift-monitoring namespaces
+
+There are several possible ways to avoid the failure:
+
+- Modify cluster pod manifests to run with `hostNetwork: true` which doesn't
+require OVN-Kubernetes CNI to create pod network, this has drawback that it
+gives cluster pod additional privileges which may not be necessarily needed.
+- Modify cluster pod manifests to select non-SmartNIC host, this may have
+issues for pods that need to run on every worker node, including the SmartNIC
+host. This also implies that additional non-SmartNIC worker nodes need be
+available
+- Find a way for the cluster pods to consume SR-IOV device by default when they
+are scheduled on the SmartNIC host. We cannot simply add SR-IOV resource request
+in cluster pod manifests, since the cluster pods are usually daemonsets or
+deployments, adding SR-IOV resource request in daemonsets or deployments will
+result in their pods be scheduled only on nodes that advertising the SR-IOV
+resources, this doesn't work for non-SmartNIC host
+- Use mutation admission controller to inject SR-IOV resource request in pod
+manifest so that pods can use SR-IOV device as their default interface.
+The problem with this approach is that admission controller runs before the
+pod is scheduled, it cannot know whether the pod is going to end up on a
+SmartNIC host or Regular OpenShift node.
+- Find a way to enable Open vSwitch for non-SR-IOV pods on SmartNIC host
+- Taint the SmartNIC hosts so that pods without corresponding toleration won't
+be scheduled to SmartNIC hosts
+
```

Comment:
probably not a stellar idea, but if we had a way for kubelet to support "default resources" for (non host network) pods it would solve the issue.

if the cluster is homogenous (all k8s workers with Smart-NICs) then the resource can be injected by default.
Subfunctions will allow better scale (currently capped at 500)